### PR TITLE
HFS/614 - Gateway implementation test cases

### PR DESCRIPTION
**Pull request by ilko-adamov**
## Link to JIRA ticket

Add a link to the JIRA ticket that the changes in this PR describe.

### *What is the problem we're trying to solve*

Add a summary of the problem and any additional context that helps explain what the issue we're trying to solve is.

### *What changes have we introduced*

Add descriptions of what changes were made to address the problem and provide a solution.

Include any links to commits that introduce significant additions of code if they help explain the process of coming to the solution. e.g Addition of test database setup, addition of first test and production class, removal of buggy code, etc.

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [ ] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [ ] Checked all code for possible refactoring
- [ ] Code pipeline builds correctly

### *Follow up actions after merging PR*

Are there any next steps that need to addressed after merging this PR? Add them here.

### CODE REVIEW

File: FinancialTransactionsApi/V1/ExceptionMiddleware.cs [hunk 1]
```diff
@@ -14,15 +14,11 @@ namespace FinancialTransactionsApi.V1
     public class ExceptionMiddleware
     {
         private readonly RequestDelegate _next;
-        private readonly ILogger<ExceptionMiddleware> _logger;
         private readonly IHostEnvironment _env;
 
-        public ExceptionMiddleware(RequestDelegate next,
-            ILogger<ExceptionMiddleware> logger,
-            IHostEnvironment env)
+        public ExceptionMiddleware(RequestDelegate next, IHostEnvironment env)
```
Comment (FaisalHackney) [inline FinancialTransactionsApi/V1/ExceptionMiddleware.cs]: Removed ILogger? Why?

File: FinancialTransactionsApi.Tests/V1/Boundary/Response/ApiResponseGenericTypeTests.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        [Fact]
+        public void ApiResponseGenericType_ReturnNull()
+        {
+            ApiResponse<ResponseWrapper<IEnumerable<Transaction>>> request = new ApiResponse<ResponseWrapper<IEnumerable<Transaction>>>();
+
+            request.Total.Should().Be(default(long));
+
+            request.Results.Should().BeNull();
```
Comment (FaisalHackney) [inline FinancialTransactionsApi.Tests/V1/Boundary/Response/ApiResponseGenericTypeTests.cs:48]: This is testing for the Results to be null.
If there are no results, should be an empty array => `results = [ ]`

So the test should be positive for 
`request.Results.Should().BeEmpty();`

File: FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs [hunk 3]
```diff
@@ -25,9 +25,9 @@ public async Task<IEnumerable<Transaction>> GetByTargetId(string targetType, Gui
         {
             var spec = new GetTransactionByTargetTypeAndTargetIdSpecification(targetType, targetId, startDate, endDate);
 
-            var response = await _databaseContext.Transactions.Where(spec.Criteria).ToListAsync().ConfigureAwait(false);
+            var response = _databaseContext.Transactions.Where(spec.Criteria);
 
-            return response?.ToDomain();
+            return await Task.FromResult(response.AsEnumerable().ToDomain()).ConfigureAwait(false);
```
Comment (FaisalHackney) [inline FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs:30]: Could the response be null here => `response?.AsEnumerable()`

**Review by FaisalHackney — CHANGES_REQUESTED** [review]

File: FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs (hunk 3 above)
Comment (ilko-adamov) [reply FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs:30]: in fact, it is not possible the response object to be null as result because null cannot say that the sequence contains no elements, it will always say that the object reference is not set to an instance of an object. So, there are two possible results: the first to return exception with message: object reference not set to an instance of an object. Second one could be an empty sequence or sequence of objects.

File: FinancialTransactionsApi.Tests/V1/Boundary/Response/ApiResponseGenericTypeTests.cs (hunk 2 above)
Comment (ilko-adamov) [reply FinancialTransactionsApi.Tests/V1/Boundary/Response/ApiResponseGenericTypeTests.cs:48]: in this test case i use a default constructor which doesn't initialize the Result property, in this situation the default value will be null, by the way this comment reminds me to create another test case for an empty collection

File: FinancialTransactionsApi/V1/ExceptionMiddleware.cs (hunk 1 above)
Comment (ilko-adamov) [reply FinancialTransactionsApi/V1/ExceptionMiddleware.cs]: Sorry, I forgot to revert the changes ... i made some tests and had to remove them, but forgot to revert the changes

Comment (ilko-adamov) [reply FinancialTransactionsApi/V1/ExceptionMiddleware.cs]: changes are reverted

**Review by FaisalHackney — APPROVED** [review]
